### PR TITLE
Provide suggestions through `rustc_confusables` annotations

### LINKED ISSUE #59450: Targeted suggestions for specific "no method" errors on standard library types
Rustc currently shows a perfectly reasonable error message when calling a nonexistent method:

```rust
use std::collections::BTreeSet;

fn main() {
    let mut set = BTreeSet::new();
    set.push("");
}
```

```console
error[E0599]: no method named `push` found for type `std::collections::BTreeSet<_>` in the current scope
 --> src/main.rs:5:9
  |
5 |     set.push("");
  |         ^^^^
```

<br>

But for specific frequently encountered cases, it would be great to show a more targeted error that gives the name of the method you almost definitely meant to call, despite not being similar enough to trigger the "did you mean" suggestion.

Candidates would be cases that arise often from refactoring (this applies to BTreeSet::push, which I see every now and then after replacing a Vec with a BTreeSet) or cases of a Rust API being named differently from an equivalent C++ API.

- BTreeSet::push ⟶ suggest `insert`
- [Vec::push_back](https://en.cppreference.com/w/cpp/container/vector/push_back) ⟶ suggest `push`
- VecDeque::push ⟶ suggest `push_back`
- Vec::length, [Vec::size](https://en.cppreference.com/w/cpp/container/vector/size) ⟶ suggest `len`

### LINKED ISSUE #87212: Add hint to correct `Vec::append` into `Vec::push` when the wrong method is used.
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a6f4c87ca812c201bfb57ef861c4d919

```rust
struct Thing;

fn main() {
    let mut stuff: Vec<Thing> = Vec::new();

    stuff.append(Thing);
}
```

The current output is:

```
error[E0308]: mismatched types
 --> src/main.rs:6:18
  |
6 |     stuff.append(Thing);
  |                  ^^^^^ expected `&mut Vec<Thing>`, found struct `Thing`
  |
  = note: expected mutable reference `&mut Vec<Thing>`
                        found struct `Thing`
```

Ideally the output should look like:

```
error[E0308]: mismatched types
 --> src/main.rs:6:18
  |
6 |     stuff.append(Thing);
  |                  ^^^^^
  |                  |
  |                  expected `&mut Vec<Thing>`, found struct `Thing`
  |                  help: use `push` to add a single element
  |
  = note: expected mutable reference `&mut Vec<Thing>`
                        found struct `Thing`
```

There should be some kind of suggestion to use `push` when a value `T` is incorrectly `append`-ed onto a `Vec<T>`. This likely only trips up newcomers to rust, but I know a fair number of python users who had this as their first error message. There are likely some some other collections this could apply to, but `Vec` would definitely constitute most of these mistakes.

If nobody is immediately opposed to this change, I would like to try to implement this. I would need some help with _where_ to make this change, but this seems like a fair introduction into editing compiler output.

It may also be helpful to just add a section mentioning `push` to the `append` function docs.

### LINKED ISSUE #108437: Using `flat_map` on an `Option<T>` should suggest using `and_then` instead
### Code

```Rust
fn main() {
    let opt = Some(1);
    opt.flat_map(|val| Some(val));
}
```

### Current output

```Shell
error[E0599]: `Option<{integer}>` is not an iterator
 --> src/main.rs:3:9
  |
3 |     opt.flat_map(|val| Some(val));
  |         ^^^^^^^^ `Option<{integer}>` is not an iterator
 --> /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/option.rs:518:1
  |
  = note: doesn't satisfy `Option<{integer}>: Iterator`
  |
  = note: the following trait bounds were not satisfied:
          `Option<{integer}>: Iterator`
          which is required by `&mut Option<{integer}>: Iterator`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` due to previous error
```

### Desired output

```Shell
error[E0599]: `Option<{integer}>` is not an iterator
 --> src/main.rs:3:9
  |
3 |     opt.flat_map(|val| Some(val));
  |         ^^^^^^^^
  |         | `Option<{integer}>` is not an iterator
  |         = help: if you wish to map and flatten the option, use `and_then`
 --> /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/option.rs:518:1
  |
  = note: doesn't satisfy `Option<{integer}>: Iterator`
  |
  = note: the following trait bounds were not satisfied:
          `Option<{integer}>: Iterator`
          which is required by `&mut Option<{integer}>: Iterator`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` due to previous error
```

### Rationale and extra context

I've run into this confusing naming issue time and time again and I think the compiler could provide a more helpful error message in this case, hinting that you should use `and_then` instead of `flat_map` if you wish to do the same operation on options (which you could write as `.map().flatten()`, hence why I always think it's `Option::flat_map` and not `Option::and_then`.)

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by estebank**
Help with common API confusion, like asking for `push` when the data structure really has `append`.

```
error[E0599]: no method named `size` found for struct `Vec<{integer}>` in the current scope
  --> $DIR/rustc_confusables_std_cases.rs:17:7
   |
LL |     x.size();
   |       ^^^^
   |
help: you might have meant to use `len`
   |
LL |     x.len();
   |       ~~~
help: there is a method with a similar name
   |
LL |     x.resize();
   |       ~~~~~~
```

Fix #59450 (we can open subsequent tickets for specific cases).

Fix #108437:

```
error[E0599]: `Option<{integer}>` is not an iterator
   --> f101.rs:3:9
    |
3   |     opt.flat_map(|val| Some(val));
    |         ^^^^^^^^ `Option<{integer}>` is not an iterator
    |
   ::: /home/gh-estebank/rust/library/core/src/option.rs:571:1
    |
571 | pub enum Option<T> {
    | ------------------ doesn't satisfy `Option<{integer}>: Iterator`
    |
    = note: the following trait bounds were not satisfied:
            `Option<{integer}>: Iterator`
            which is required by `&mut Option<{integer}>: Iterator`
help: you might have meant to use `and_then`
    |
3   |     opt.and_then(|val| Some(val));
    |         ~~~~~~~~
```

On type error of method call arguments, look at confusables for suggestion. Fix #87212:

```
error[E0308]: mismatched types
    --> f101.rs:8:18
     |
8    |     stuff.append(Thing);
     |           ------ ^^^^^ expected `&mut Vec<Thing>`, found `Thing`
     |           |
     |           arguments to this method are incorrect
     |
     = note: expected mutable reference `&mut Vec<Thing>`
                           found struct `Thing`
note: method defined here
    --> /home/gh-estebank/rust/library/alloc/src/vec/mod.rs:2025:12
     |
2025 |     pub fn append(&mut self, other: &mut Self) {
     |            ^^^^^^
help: you might have meant to use `push`
     |
8    |     stuff.push(Thing);
     |           ~~~~
```

### CODE REVIEW

**Comment by chenyukang** [comment]
Fixes #87212

**Comment by chenyukang** [comment]
can you also add a annotation for #108437

**Comment by estebank** [comment]
@chenyukang added :)

**Review by oli-obk — CHANGES_REQUESTED** [review]
whoops started reviewing this and got distracted. Will do another round tomorrow

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs [hunk 1]
```diff
@@ -563,80 +563,73 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            //  - methods with short levenshtein distance
+
+            // Look for commonly confusable method names considering arguments.
+            if let Some(name) = self.confusable_method_name(
+                err,
+                callee_ty.peel_refs(),
+                call_name,
+                Some(input_types.clone()),
+            ) {
+                return Some(name);
```
Comment (oli-obk) [inline compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:582]: Afaict the return value is never used

**Comment by nnethercote** [comment]
For posterity: this caused a large performance regression in certain cases where errors were being emitted. I had an example involving a refactoring of rustc itself where I had hundreds of errors and each error was taking ~5 seconds to emit(!)  #125100 fixes the problem.

**Pull request merged**